include/ngp_utils/NgpLoopUtils.h

```diff
@@ -0,0 +1,401 @@
+ *
+ * @param sz The number of STK buckets we are looping over
+ * @param bytes_per_team Bytes to be allocated at team parallelism level
+ * @param bytes_per_thread Bytes to be allocated at thread level
+ * @return Kokkos TeamPolicy instance for use with Kokkos::parallel_for
+ */
+template <typename TeamPolicy>
+inline TeamPolicy
+ngp_mesh_team_policy(
+  const size_t sz,
+  const size_t bytes_per_team,
+  const size_t bytes_per_thread)
+{
+  TeamPolicy policy(sz, Kokkos::AUTO);
+  return policy.set_scratch_size(
+    1, Kokkos::PerTeam(bytes_per_team), Kokkos::PerThread(bytes_per_thread));
+}
+
+/** Estimate the bytes required per thread to store ScratchViews for
+ *  element data.
+ *
+ *  @param ndim Spatial dimension
+ *  @param dataReq Element data requests object
+ *  @return bytes_per_thread
+ */
+template<typename T, typename DataReqType>
+inline int ngp_calc_thread_shmem_size(int ndim, const DataReqType& dataReq)
+{
+  int preReqSize = get_num_bytes_pre_req_data<T>(dataReq, ndim);
+  int mdvSize = MultiDimViews<T>::bytes_needed(
+    dataReq.get_total_num_fields(), count_needed_field_views(dataReq));
+
+#ifndef KOKKOS_ENABLE_CUDA
+  // On host account for extra data to store the SIMD and non-SIMD versions of
+  // the ScratchViews
+  return (preReqSize + mdvSize) * 2 * simdLen;
+#else
+  return (preReqSize + mdvSize);
+#endif
+}
+
+/** Estimate the bytes required per thread for face/element ScratchViews
+ *
+ *  @param ndim Spatial dimension
+ *  @param Face data requests object
+ *  @param Element data requests object (this is the parent element of the face)
+ *  @param MasterElementInfo object that contains nodesPerElem/nodesPerFace etc.
+ *  @return bytes_per_thread
+ */
+template<typename T, typename DataReqType>
+inline int ngp_calc_thread_shmem_size(
+  int ndim,
+  const DataReqType& faceDataReq,
+  const DataReqType& elemDataReq)
+{
+  const int faceMemSize = ngp_calc_thread_shmem_size<T>(ndim, faceDataReq);
+  const int elemMemSize = ngp_calc_thread_shmem_size<T>(ndim, elemDataReq);
+
+  return (faceMemSize + elemMemSize);
+}
+
+/** Get the nodes per element from the MasterElement registered
+ *
+ *  We need a convenience function because the ME instance is created on the device.
+ */
+template<typename DataReqType>
+inline int nodes_per_element(const DataReqType& dataReqIn)
+{
+  int npe = 0;
+  DataReqType dataReq(dataReqIn);
+  Kokkos::parallel_reduce(
+    1, KOKKOS_LAMBDA(int, int& n) {
+      auto* meSCS = dataReq.get_cvfem_surface_me();
+      auto* meSCV = dataReq.get_cvfem_volume_me();
+      n = ((meSCS != nullptr) ? meSCS->nodesPerElement_ :
+           (meSCV != nullptr) ? meSCV->nodesPerElement_ : 0);
+    }, npe);
+
+  NGP_ThrowRequire(npe != 0);
+  return npe;
+}
+
+/** Get the nodes per face from the MasterElement registered
+ *
+ *  We need a convenience function because the ME instance is created on the device.
+ */
+template<typename DataReqType>
+inline int nodes_per_face(const DataReqType& dataReqIn)
```

Comment:
@sayerhs @overfelt I think we need one of these for num_integration_points to fix the issue at line 56 in AssembleFaceElemSolverAlgorithm.h